Network/Socket/Safe.hs

```diff
@@ -0,0 +1,78 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Network.Socket.Safe
+-- Copyright   :  Echo Nolan 2016
+-- License     :  BSD-style (see the file libraries/network/LICENSE)
+--
+-- Maintainer  :  libraries@haskell.org
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- A drop in replacement for "Network.Socket" that sacrifices some performance
+-- for correctness. Specifically, this module's functions check that a socket
+-- hasn't been closed before attempting to read or write from it. With the
+-- "Network.Socket" API, reading or writing to a socket after closing it causes
+-- undefined behavior. In this module it always throws an exception. N.b. this
+-- serializes all use of a given socket via an MVar.
+--
+-- See "Network.Socket" for API documentation.
+-----------------------------------------------------------------------------
+
+-- We reexport things with attached warnings, but we don't want those warnings
+-- here.
+{-# OPTIONS_GHC -fno-warn-warnings-deprecations #-}
+
+module Network.Socket.Safe
+  (
+    module Network.Socket
+  , send
+  , sendTo
+  , recv
+  , recvFrom
+  , recvLen
+  , sendBuf
+  , recvBuf
+  , sendBufTo
+  , recvBufFrom
+  ) where
+
+import Network.Socket hiding (send, sendTo, recv, recvFrom, recvLen, sendBuf,
+                              recvBuf, sendBufTo, recvBufFrom)
+import qualified Network.Socket as Unsafe
+
+import Network.Socket.Internal
+
+import Data.Word (Word8)
+import Foreign.Ptr (Ptr)
+
+{-# WARNING send "Use send defined in \"Network.Socket.ByteString.Safe\"" #-}
```

Comment:
Does it make sense to introduce new `String` variants that immediately have a warning? If we choose to introduce these then they most definitely must have a warning. The alternative is to not provide them in the first place.

I assume the argument for providing them is that a new-comer might be confused and frustrated when they don't find them. The converse of this is that the existing `String` functions are already well documented as being obsolete and introducing new `String` functions could cause more confusion and a maintenance burden.

I'd rather not continue to contribute to the "`String` problem", so I'd vote against these variants.